[JENKINS-75799] Replace AbstractLazyLoadRunMap.index by CopyOnWriteMap

dukhlov opened the pull request:
See [JENKINS-75799](https://issues.jenkins.io/browse/JENKINS-75799)

### Background
We have Jenkins cluster which has a large number of jobs and build history. To make it working we use
`-Djenkins.model.lazy.BuildReference.MODE=weak`. It works but, but my further investigation shows
that RunMap implementation can be optimised and simplified.


### Problem
Jenkins Job's RanMap is based on `index` (`TreeMap` which initially empty but eventually keeps full set of BuildReferences) + `numberOnDisk` (`SortedList` which initialised with full set of existing builds at Job loading)

As far as I understood this is "historical" implementation this solved lazy loading problem when index values were build objects (not BuildRefferences to the build object as we have now) and can be simplified to provide ability for further RunMap optimisation

### Solution
Use a single TreeMap which initialised with full set of existing builds at Job loading (with BuildReference as a value which delays build objects loading from the disk)

### Future Work
Optimise RunMap lookup methods and Build objects lazy loading to load object only if we need it's value

### Bonus
Decrease memory usage a bit by removing `numberOnDisk` SortedList, simplify and make more compact code base under maintenance.

### Testing done

This part of code covered by unit and integration tests running during PR check. + Manual testing on local environment done

### Proposed changelog entries

- Simplify underlining RunMap data structure

### Proposed changelog category

/label internal

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers
@jenkinsci/core-pr-reviewers

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

Commit:
JENKINS-75799: Replace AbstractLazyLoadRunMap.index by hudson.util.CopyOnWriteMap

jglick commented on core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java:
```diff
@@ -250,40 +243,6 @@ public boolean contains(Object v) {
     }
 
-    /**
-     * Historical holder for map.
-     *
-     * TODO all this mess including {@link #numberOnDisk} could probably be simplified to a single {@code TreeMap<Integer,BuildReference<R>>}
-     * where a null value means not yet loaded and a broken entry just uses {@code NoHolder}.
```
#1379 FTR

jglick commented:
This change regresses a functional test in CloudBees CI (`MultiBranchProjectTest.deadBranch` FTR). I am looking into what specifically broke and whether it constitutes a genuine regression or just an issue with the test, and how it might be reproduced in an OSS test case.

jglick commented on core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java:
```diff
@@ -657,70 +612,30 @@ public synchronized void putAll(Map<? extends Integer, ? extends R> rhs) {
-        }
-        return index.byNumber;
+    /*package*/ SortedMap<Integer, BuildReference<R>> all() {
+        return core;
```
This change invalidates the Javadoc for the method.

dukhlov commented on core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java:
```diff
@@ -616,36 +590,17 @@ protected R _put(R value) {
-    private R unwrap(BuildReference<R> ref) {
-        return ref != null ? ref.get() : null;
+        BuildReference<R> old = core.put(n, createReference(r));
+        return resolveBuildRef(old);
```
@jglick, hm... I understand the problem, but suggested change breakes put method semantic which expected to return previous value. old.get() can return null if currently unused Run GCed (if we use weak BuildReference holder), or if is wasn't loaded yet. look like better solution. Looks like right solution is adding another internal package local method ('replace' for instance) which won't return previous value, or use putAll() method which does not return previous value... Or in your opinion it is better just to keep previous behaviour in this place?

jglick replied:
> adding another internal package local method (`replace` for instance) which won't return previous value

That would work.

dukhlov replied:
done

Commit:
JENKINS-75799: Fix reloading current building builds during Job.onLoad

Use single putAll method for RunMap initialization instead of lots of put call to avoid
  * potential loading old values from the disk
  * multiple (per each 'put') data copying (caused by using CopyOnWriteMap)

Commit:
JENKINS-75799: Fix AbstractLazyLoadRunMap.all() javadoc comment

jglick commented:
> please do not force-push within a PR branch

Repeating this request.

jglick commented on core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java:
```diff
@@ -522,7 +483,7 @@ public boolean runExists(int number) {
-            for (int m : numberOnDisk) {
+            for (int m : core.descendingMap().keySet()) {
                 if (m < n) {
                     // TODO could be made more efficient with numberOnDisk.find
```
Obsolete I think.

```suggestion
```

dukhlov commented:
> > please do not force-push within a PR branch
> 
> Repeating this request.

hm, I just rebased on current master and didn't change commit set. Is it not desired as well?

dukhlov committed:
Apply suggestion from @jglick

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

dukhlov committed:
Update core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

jglick commented:
> rebased on current master

No need, just click **Update branch** or generally merge in upstream `master`.

Commit:
JENKINS-75799: Ignoring whitespaces

dukhlov committed:
Merge branch 'master' into refactor-lazy-loading

jglick reviewed (approved):
After the resolution of https://github.com/jenkinsci/jenkins/pull/10759#issuecomment-3046122083 in 1b128a6793acd46a4f4e141a9facababbf73b25c this seems to pass all tests in CloudBees CI. The code being touched is quite complex and the change to it is substantial (albeit making it somewhat simpler) and this is central to the application’s functioning, so help reviewing this would really be appreciated.

timja reviewed (approved):
LGTM

jglick commented:
I have already reviewed it. I do not know what the process is from there.

dukhlov commented:
> I have already reviewed it. I do not know what the process is from there.

Ah, got it. 

@krisstern, perhaps you can help. I am not familiar with Jenkins dev process, 3 weeks passed, and I don't understand if the process is stuck or not.

A1exKH reviewed (approved):
LGTM.
